src/gar.lua

```diff
@@ -1577,4 +1577,271 @@ function gar.getPaginatedDelegations(address, cursor, limit, sortBy, sortOrder)
+
+	-- Assert source has enough stake to redelegate and remove the stake from the source
+	if delegateAddress == sourceAddress then
+		-- check if he gateway can afford to redelegate from itself
+
+		if vaultId then
+			-- Get the redelegation amount from the operator vault
+
+			local existingVault = sourceGateway.vaults[vaultId]
+			assert(existingVault, "Vault not found on the operator.")
+			assert(
+				existingVault.balance >= stakeToTakeFromSource,
+				"Quantity must be less than or equal to the vaulted stake amount."
+			)
+
+			if existingVault.balance == stakeToTakeFromSource then
+				-- The operator vault has been emptied
+				sourceGateway.vaults[vaultId] = nil
+			else
+				-- The operator vault has been partially emptied
+				sourceGateway.vaults[delegateAddress][vaultId].balance = sourceGateway.vaults[delegateAddress][vaultId].balance
+					- stakeToTakeFromSource
+			end
+		else
+			-- Get the redelegation amount from the operator stakes
+			local maxWithdraw = sourceGateway.operatorStake - gar.getSettings().operators.minStake
+			assert(
+				stakeToTakeFromSource <= maxWithdraw,
+				"Resulting stake is not enough to maintain the minimum operator stake of "
+					.. gar.getSettings().operators.minStake
+					.. " IO"
+			)
+
+			sourceGateway.operatorStake = sourceGateway.operatorStake - stakeToTakeFromSource
+		end
+	else
+		local existingDelegate = sourceGateway.delegates[delegateAddress]
+		assert(existingDelegate, "This delegate has no stake to redelegate.")
+
+		if vaultId then
+			local existingVault = existingDelegate.vaults[vaultId]
+			assert(existingVault, "Vault not found on the delegate.")
+			assert(
+				existingVault.balance >= stakeToTakeFromSource,
+				"Quantity must be less than or equal to the vaulted stake amount."
+			)
+
+			if existingVault.balance == stakeToTakeFromSource then
+				-- The vault has been emptied
+				sourceGateway.delegates[delegateAddress].vaults[vaultId] = nil
+
+				if
+					-- If all vaults and stake are emptied, remove the delegate
+					sourceGateway.delegates[delegateAddress].delegatedStake == 0
+					and next(sourceGateway.delegates[delegateAddress].vaults) == nil
+				then
+					gar.pruneDelegateFromGateway(delegateAddress, sourceGateway)
+				end
+			else
+				-- The vault has been partially emptied
+				sourceGateway.delegates[delegateAddress].vaults[vaultId].balance = sourceGateway.delegates[delegateAddress].vaults[vaultId].balance
+					- stakeToTakeFromSource
+			end
+		else
+			-- Check if the delegate has enough stake to redelegate
+			assert(
+				existingDelegate.delegatedStake >= stakeToTakeFromSource,
+				"Quantity must be less than or equal to the delegated stake amount."
+			)
+
+			-- Check if the delegate will have enough stake left after re-delegating
+			local existingStake = existingDelegate.delegatedStake
+			local requiredMinimumStake = sourceGateway.settings.minDelegatedStake
+			local maxAllowedToWithdraw = existingStake - requiredMinimumStake
+			assert(
```

Comment:
It's interesting that we have this barrier here since the user could always commence a withdraw of their stake and then begin a redelegation. We grapple with this a bit with funding plans where, since we acknowledge this potential user workaround, we err on the side of a good UX for the user and simply "break a 20" on their minimum stake if necessary. We'll use it all if we have to, or we'll leave behind the excess in a withdraw vault. I'm wondering why we wouldn't just do the same here (we can check with Jon).